### PR TITLE
Jvmdiscovery improvements #125

### LINKED ISSUE #125: Improvement to the JVM discovery tab
Hi, I would to push a contribution to improve the integration for JVM discovery. 
The current UI has visual defects and behavioral glitch due to regression introduced in libs upgrades.

Inbound PR fixes most of these issues and drop legacy code from the template.
image

If/once we agree on this improvement, I will follow up shortly with another to provide a strategy pattern for discovery, as the Jolokia discovery agent is not cutting it in most modern scenarios.

### PR DESCRIPTION
**Pull request by yuhusolutions**
Hi, please review kindly review the following. 

Fixes #125

### CODE REVIEW

File: plugins/jvm/ts/discover.ts [hunk 1]
```diff
@@ -4,28 +4,31 @@ namespace JVM {
...
-      (<any>$)($event.currentTarget).parents('.popover').prev().popover('hide');
+      (<any>$)($event.currentTarget).parents('.popover').hide();
     };
 
     function getMoreJvmDetails(agents) {
       for (let key in agents) {
         const agent = agents[key];
         if (agent.url && !agent.secured) {
-          const dedicatedJolokia = createJolokia(Core.useProxyIfExternal(agent.url), agent.username, agent.password);
+          const dedicatedJolokia = createJolokia(agent.url, agent.username, agent.password);
```
Comment (tadayosi) [inline plugins/jvm/ts/discover.ts:29]: Why is usage of `Core.useProxyIfExternal(..)` removed?

File: plugins/jvm/ts/discover.ts [hunk 2]
```diff
@@ -4,28 +4,31 @@ namespace JVM {
...
           agent.startTime = dedicatedJolokia.getAttribute('java.lang:type=Runtime', 'StartTime');
-          if (!$scope.hasName(agent)) {//only look for command if agent vm is not known
-            agent.command = dedicatedJolokia.getAttribute('java.lang:type=Runtime', 'SystemProperties', 'sun.java.command');
-          }
+          agent.command = dedicatedJolokia.getAttribute('java.lang:type=Runtime', 'SystemProperties', 'sun.java.command');
```
Comment (tadayosi) [inline plugins/jvm/ts/discover.ts:31]: Why is `$scope.hasName(agent)` check removed?

**Review by tadayosi — CHANGES_REQUESTED** [review]

**Comment by tadayosi** [comment]
Thanks, but I need more clarifications before merging it. See my comments.

File: plugins/jvm/ts/discover.ts (hunk 2 above)
Comment (yuhusolutions) [reply plugins/jvm/ts/discover.ts:31]: Hi, I found an inconsistency in the UI where the java command would not be printed when mocking server detail.
The if clause is "arbitrary" and the comment doesn't help understanding the logic of the author behind this. Maybe you can shed light  on why shouldn't we look for the java command when agent VM is known? 

From a UI point of view, this leads to element in the list not displaying the same information which looks like a defect at first glance.

File: plugins/jvm/ts/discover.ts (hunk 1 above)
Comment (yuhusolutions) [reply plugins/jvm/ts/discover.ts:29]: Good catch, that wasn't meant to be pushed! Will revert this.

Comment (yuhusolutions) [reply plugins/jvm/ts/discover.ts]: I reverted it ,  but then I noticed that the function useProxyIfExternal strips the protocol when building the proxy URL.

It yields a HTTP 406 with following URL: 
`http://127.0.0.1:8080/integration/proxy/localhost/8084/actuator/jolokia/read/java.lang%3Atype%3DRuntime/SystemProperties/sun.java.command?ignoreErrors=true&canonicalNaming=false`

Where we should expect the protocol *http* to be part of the path:
`http://127.0.0.1:8080/integration/proxy/http/localhost/8084/actuator/jolokia/read/java.lang:type=Runtime/SystemProperties/sun.java.command?ignoreErrors=true&canonicalNaming=false`

I don't want to toy around the Hawtio core functions, do you have any idea why the protool is being stripped ?


```js
            // hawtio-core.js line 4379
            // lets remove the http stuff
            var idx = connectUrl.indexOf("://");
            if (idx > 0) {
                connectUrl = connectUrl.substring(idx + 3);
            }

```
image

image

Comment (yuhusolutions) [reply plugins/jvm/ts/discover.ts]: @tadayosi see last commit. I created a new function `toJolokiaProxyUrl` to work around the glitch.
This now assumes that everything will go through the proxy, which I believe is fine. 
Let me know if this solution is acceptable.

File: plugins/jvm/ts/discover.ts (hunk 2 above)
Comment (yuhusolutions) [reply plugins/jvm/ts/discover.ts:45]: @tadayosi  see above comment.

**Comment by yuhusolutions** [comment]
Hi @tadayosi, I haven't heard back you, I'm closing the the PR assuming the discovery plugin is abandoned.

**Pull request closed without merging**